docs/RFCS/grpc.md

```diff
@@ -0,0 +1,75 @@
+- Feature Name: grpc
+- Status: draft
+- Start Date: 2015-12-07
+- RFC PR:
+- Cockroach Issue: #2381, #3013, #3421
+
+# Summary
+
+Replace the custom protocol in `cockroach/rpc/codec` with GRPC. This
+will affect both internal communication and potential client-facing
+APIs (assuming we offer other public APIs besides `pgwire`). GRPC was
+previously discussed in #2381 for client-facing APIs; this RFC extends
+the proposal to internal RPCs as well.
+
+# Motivation
+
+The primary motivation is to minimize the impact that raft snapshots
+have on other RPCs (#3013). Our RPC codec transmits each request as a
+single chunk on the network, which can block other requests for a
+significant period of time (leading to more severe consequences if
+raft heartbeats or leader lease operations are blocked for too long).
+Since GRPC is based on HTTP/2, it has built-in support for
+multiplexing large messages.
+
+Secondary benefits of GRPC include support for bidirectional streaming
+channels, which are a good fit for raft and gossip messages that do
+not fit the "one request, one response" pattern, and may be useful for
+future distributed SQL workloads. There are also benefits to adopting
+a more widely-used protocol instead of our current custom one.
+
+# Detailed design
+
+Remove all use of `cockroach/rpc/codec` and `net/rpc`, replacing them
+with GRPC. `cockroach/rpc` be removed, or may become a thin wrapper
+around GRPC. In either case, interfaces will need to be extended to
+allow streaming; it probably makes more sense to pass GRPC objects
+around everywhere than to attempt to cover them with our own
+abstraction.
+
+# Drawbacks
+
+GRPC is currently slower than our RPC codec ([benchmark results](
+https://github.com/cockroachdb/rpc-bench)). It
```

Comment:
Yes, it's because streaming RPCs don't include headers with every message, so it's just one Write syscall per message. I haven't benchmarked it but on paper streaming mode should be pretty close to our current system. I don't think we can make everything use the streaming mode but it's a good fit for at least raft and gossip.